// @ant-design/plots/lib/index.js
"use strict";
var __createBinding = (this && this.__createBinding) || (Object.create ? (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    var desc = Object.getOwnPropertyDescriptor(m, k);
    if (!desc || ("get" in desc ? !m.__esModule : desc.writable || desc.configurable)) {
      desc = { enumerable: true, get: function() { return m[k]; } };
    }
    Object.defineProperty(o, k2, desc);
}) : (function(o, m, k, k2) {
    if (k2 === undefined) k2 = k;
    o[k2] = m[k];
}));
var __exportStar = (this && this.__exportStar) || function(m, exports) {
    for (var p in m) if (p !== "default" && !Object.prototype.hasOwnProperty.call(exports, p)) __createBinding(exports, m, p);
};
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.adaptors = exports.measureTextWidth = exports.flow = exports.FUNNEL_CONVERSATION_FIELD = exports.getCanvasPattern = exports.G2 = exports.BidirectionalBar = exports.Mix = exports.MultiView = exports.Venn = exports.Violin = exports.Treemap = exports.Sankey = exports.Gauge = exports.CirclePacking = exports.RadialBar = exports.Stock = exports.DualAxes = exports.Sunburst = exports.WordCloud = exports.Waterfall = exports.TinyLine = exports.TinyColumn = exports.TinyArea = exports.Scatter = exports.Chord = exports.Rose = exports.RingProgress = exports.Facet = exports.Radar = exports.Progress = exports.Pie = exports.Heatmap = exports.Liquid = exports.Line = exports.Histogram = exports.Funnel = exports.Column = exports.Bullet = exports.Box = exports.Bar = exports.Area = void 0;
var area_1 = __importDefault(require("./components/area"));
exports.Area = area_1.default;
var bar_1 = __importDefault(require("./components/bar"));
exports.Bar = bar_1.default;
var bullet_1 = __importDefault(require("./components/bullet"));
exports.Bullet = bullet_1.default;
var column_1 = __importDefault(require("./components/column"));
exports.Column = column_1.default;
var funnel_1 = __importDefault(require("./components/funnel"));
exports.Funnel = funnel_1.default;
var histogram_1 = __importDefault(require("./components/histogram"));
exports.Histogram = histogram_1.default;
var line_1 = __importDefault(require("./components/line"));
exports.Line = line_1.default;
var box_1 = __importDefault(require("./components/box"));
exports.Box = box_1.default;
var liquid_1 = __importDefault(require("./components/liquid"));
exports.Liquid = liquid_1.default;
var heatmap_1 = __importDefault(require("./components/heatmap"));
exports.Heatmap = heatmap_1.default;
var pie_1 = __importDefault(require("./components/pie"));
exports.Pie = pie_1.default;
var gauge_1 = __importDefault(require("./components/gauge"));
exports.Gauge = gauge_1.default;
var progress_1 = __importDefault(require("./components/progress"));
exports.Progress = progress_1.default;
var radar_1 = __importDefault(require("./components/radar"));
exports.Radar = radar_1.default;
var ring_progress_1 = __importDefault(require("./components/ring-progress"));
exports.RingProgress = ring_progress_1.default;
var rose_1 = __importDefault(require("./components/rose"));
exports.Rose = rose_1.default;
var chord_1 = __importDefault(require("./components/chord"));
exports.Chord = chord_1.default;
var scatter_1 = __importDefault(require("./components/scatter"));
exports.Scatter = scatter_1.default;
var tiny_area_1 = __importDefault(require("./components/tiny-area"));
exports.TinyArea = tiny_area_1.default;
var tiny_column_1 = __importDefault(require("./components/tiny-column"));
exports.TinyColumn = tiny_column_1.default;
var dual_axes_1 = __importDefault(require("./components/dual-axes"));
exports.DualAxes = dual_axes_1.default;
var tiny_line_1 = __importDefault(require("./components/tiny-line"));
exports.TinyLine = tiny_line_1.default;
var waterfall_1 = __importDefault(require("./components/waterfall"));
exports.Waterfall = waterfall_1.default;
var word_cloud_1 = __importDefault(require("./components/word-cloud"));
exports.WordCloud = word_cloud_1.default;
var sunburst_1 = __importDefault(require("./components/sunburst"));
exports.Sunburst = sunburst_1.default;
var stock_1 = __importDefault(require("./components/stock"));
exports.Stock = stock_1.default;
var radial_bar_1 = __importDefault(require("./components/radial-bar"));
exports.RadialBar = radial_bar_1.default;
var sankey_1 = __importDefault(require("./components/sankey"));
exports.Sankey = sankey_1.default;
var treemap_1 = __importDefault(require("./components/treemap"));
exports.Treemap = treemap_1.default;
var violin_1 = __importDefault(require("./components/violin"));
exports.Violin = violin_1.default;
var venn_1 = __importDefault(require("./components/venn"));
exports.Venn = venn_1.default;
var circle_packing_1 = __importDefault(require("./components/circle-packing"));
exports.CirclePacking = circle_packing_1.default;
var facet_1 = __importDefault(require("./components/facet"));
exports.Facet = facet_1.default;
var mix_1 = __importDefault(require("./components/mix")); // 和 Mix 相同
exports.MultiView = mix_1.default;
var mix_2 = __importDefault(require("./components/mix"));
exports.Mix = mix_2.default;
var bidirectional_bar_1 = __importDefault(require("./components/bidirectional-bar"));
exports.BidirectionalBar = bidirectional_bar_1.default;
var g2plot_1 = require("@antv/g2plot");
Object.defineProperty(exports, "G2", { enumerable: true, get: function () { return g2plot_1.G2; } });
Object.defineProperty(exports, "flow", { enumerable: true, get: function () { return g2plot_1.flow; } });
Object.defineProperty(exports, "measureTextWidth", { enumerable: true, get: function () { return g2plot_1.measureTextWidth; } });
Object.defineProperty(exports, "adaptors", { enumerable: true, get: function () { return g2plot_1.adaptors; } });
Object.defineProperty(exports, "getCanvasPattern", { enumerable: true, get: function () { return g2plot_1.getCanvasPattern; } });
Object.defineProperty(exports, "FUNNEL_CONVERSATION_FIELD", { enumerable: true, get: function () { return g2plot_1.FUNNEL_CONVERSATION_FIELD; } });
exports.default = {
    Area: area_1.default,
    Bar: bar_1.default,
    Box: box_1.default,
    Bullet: bullet_1.default,
    Column: column_1.default,
    Funnel: funnel_1.default,
    Histogram: histogram_1.default,
    Line: line_1.default,
    Liquid: liquid_1.default,
    Heatmap: heatmap_1.default,
    Pie: pie_1.default,
    Progress: progress_1.default,
    Radar: radar_1.default,
    Facet: facet_1.default,
    RingProgress: ring_progress_1.default,
    Rose: rose_1.default,
    Chord: chord_1.default,
    Scatter: scatter_1.default,
    TinyArea: tiny_area_1.default,
    TinyColumn: tiny_column_1.default,
    TinyLine: tiny_line_1.default,
    Waterfall: waterfall_1.default,
    WordCloud: word_cloud_1.default,
    Sunburst: sunburst_1.default,
    DualAxes: dual_axes_1.default,
    Stock: stock_1.default,
    RadialBar: radial_bar_1.default,
    Gauge: gauge_1.default,
    CirclePacking: circle_packing_1.default,
    Sankey: sankey_1.default,
    Treemap: treemap_1.default,
    Violin: violin_1.default,
    Venn: venn_1.default,
    MultiView: mix_1.default,
    Mix: mix_2.default,
    BidirectionalBar: bidirectional_bar_1.default,
    // 直接导出 G2Plot 相关方法
    getCanvasPattern: g2plot_1.getCanvasPattern,
    FUNNEL_CONVERSATION_FIELD: g2plot_1.FUNNEL_CONVERSATION_FIELD,
    G2: g2plot_1.G2,
    flow: g2plot_1.flow,
    measureTextWidth: g2plot_1.measureTextWidth,
    adaptors: g2plot_1.adaptors,
};
__exportStar(require("./interface"), exports);
