// @ant-design/plots/lib/utils/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.unmount = exports.render = exports.getChart = exports.uuid = exports.deepClone = exports.setPath = exports.hasPath = exports.clone = exports.isType = void 0;
var utils_1 = require("./utils");
Object.defineProperty(exports, "isType", { enumerable: true, get: function () { return utils_1.isType; } });
Object.defineProperty(exports, "clone", { enumerable: true, get: function () { return utils_1.clone; } });
Object.defineProperty(exports, "hasPath", { enumerable: true, get: function () { return utils_1.hasPath; } });
Object.defineProperty(exports, "setPath", { enumerable: true, get: function () { return utils_1.setPath; } });
Object.defineProperty(exports, "deepClone", { enumerable: true, get: function () { return utils_1.deepClone; } });
Object.defineProperty(exports, "uuid", { enumerable: true, get: function () { return utils_1.uuid; } });
var getChart_1 = require("./getChart");
Object.defineProperty(exports, "getChart", { enumerable: true, get: function () { return getChart_1.getChart; } });
var render_1 = require("./render");
Object.defineProperty(exports, "render", { enumerable: true, get: function () { return render_1.render; } });
Object.defineProperty(exports, "unmount", { enumerable: true, get: function () { return render_1.unmount; } });
